src/xcode/ENA/ENA/Source/Services/HealthCertificateValidationService/HealthCertificateValidationService.swift

```diff
@@ -172,4 +212,557 @@ final class HealthCertificateValidationService: HealthCertificateValidationProvi
+			completion(.failure(.INVALIDATION_RULE_SERVER_ERROR))
+		}
+	}
+	
+	private func invalidationRules(_ data: Data, completion: (Result<[Rule], HealthCertificateValidationError>) -> Void) {
+		let extractInvalidationRulesResult = ValidationRulesAccess().extractValidationRules(from: data)
+		
+		switch extractInvalidationRulesResult {
+		case let .success(invalidationRules):
+			completion(.success(invalidationRules))
+		case .failure:
+			completion(.failure(.INVALIDATION_RULE_JSON_DECODING_FAILED))
+		}
+	}
+	
+	// MARK: - External rule parameters
+	
+	private func assembleAcceptanceExternalRuleParameters(
+		healthCertificate: HealthCertificate,
+		arrivalCountry: String,
+		validationClock: Date,
+		valueSet: SAP_Internal_Dgc_ValueSets
+	) -> ExternalParameter {
+		
+		let mappedValueSets = mapValueSetsForExternalParameter(valueSet: valueSet)
+		let mappedCertificateType = mapForExternalParameter(healthCertificate.type)
+		
+		let externalRuleParameter = ExternalParameter(
+			validationClock: validationClock,
+			valueSets: mappedValueSets,
+			countryCode: arrivalCountry,
+			exp: mapUnixTimestampsInSecondsToDate(healthCertificate.cborWebTokenHeader.expirationTime),
+			iat: mapUnixTimestampsInSecondsToDate(healthCertificate.cborWebTokenHeader.issuedAt),
+			certificationType: mappedCertificateType,
+			issueCountryCode: healthCertificate.cborWebTokenHeader.issuer
+		)
+		
+		return externalRuleParameter
+	}
+	
+	private func assembleInvalidationExternalRuleParameters(
+		healthCertificate: HealthCertificate,
+		arrivalCountry: String,
+		validationClock: Date,
+		valueSet: SAP_Internal_Dgc_ValueSets
+	) -> ExternalParameter {
+		
+		let mappedValueSets = mapValueSetsForExternalParameter(valueSet: valueSet)
+		let mappedCertificateType = mapForExternalParameter(healthCertificate.type)
+
+		let externalRuleParameter = ExternalParameter(
+			validationClock: validationClock,
+			valueSets: mappedValueSets,
+			countryCode: healthCertificate.cborWebTokenHeader.issuer,
+			exp: mapUnixTimestampsInSecondsToDate(healthCertificate.cborWebTokenHeader.expirationTime),
+			iat: mapUnixTimestampsInSecondsToDate(healthCertificate.cborWebTokenHeader.issuedAt),
+			certificationType: mappedCertificateType,
+			issueCountryCode: healthCertificate.cborWebTokenHeader.issuer
+		)
+		
+		return externalRuleParameter
+	}
+	
+	/// Maps our valueSet on the value set of CertLogic. See https://github.com/corona-warn-app/cwa-app-tech-spec/blob/proposal/business-rules-dcc/docs/spec/dgc-validation-rules-client.md#data-structure-of-external-rule-parameters
+	private func mapValueSetsForExternalParameter(
+		valueSet: SAP_Internal_Dgc_ValueSets
+	) -> [String: [String]] {
+		var dictionary = [String: [String]]()
+		dictionary["country-2-codes"] = allCountryCodes
+		dictionary["covid-19-lab-result"] = valueSet.tcTr.items.map { $0.key }
+		dictionary["covid-19-lab-test-manufacturer-and-name"] = valueSet.ma.items.map { $0.key }
```

Comment:
```suggestion
		dictionary["covid-19-lab-test-manufacturer-and-name"] = valueSet.tcMa.items.map { $0.key }
```
I didn't check, but I think this is probably the correct one. We already had a bug because I mixed them up once 😅

Reply:
Uh good point, this I did not know. Thx!